modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+    if _sanityChecks then
+      assert(expr);
+  }
+
+  //
+  // Basically just a wrapper around _ddata at this point, with one based
+  // indexing to avoid logical errors.
+  //
+  pragma "no doc"
+  class ListBlock {
+
+    type _etype;
+    var _data: _ddata(_etype);
+    const capacity: int;
+
+    proc init(type _etype, capacity: int) {
+      _sanity(capacity > 0);
+      this._etype = _etype;
+      //
+      // Use `c_calloc` to avoid calling initializers. TODO: Is use of calloc
+      // safe in a multi-locale context? Michael said that the cast to
+      // _ddata will convert the c_ptr to a wide pointer.
+      //
+      this._data = c_calloc(_etype, capacity):_ddata(_etype);
+      this.capacity = capacity;
+    }
+
+    proc deinit() {
+      c_free(_data:c_void_ptr);
+    }
+
+    inline proc this(i: int) ref {
+      _sanity(i >= 1 && i <= capacity);
+      return _data[i - 1];
+    }
+  }
+
+  //
+  // We can change the lock type later. Use a spinlock for now, even if it
+  // is suboptimal in cases where long critical sections have high
+  // contention (IE, lots of tasks trying to insert into the middle of this
+  // List, or any operation that is O(n)).
+  //
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  //
+  // Use a wrapper class to let the List class have a const ref receiver even
+  // when `parSafe` is `true` and the List lock is used.
+  //
+  pragma "no doc"
+  class LockWrapper {
+    
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  /*
+    A List is a lightweight container suitable for building up and iterating
+    over a collection of elements in a structured manner. It is intended to
+    replace the use of arrays to perform "vector-like" operations. Unlike a
+    stack, the List type also supports inserts or removals into the middle of
+    the List. The List type is close in spirit to its Python counterpart, with
+    fast O(1) random access and append operations.
+
+    The List type is not parallel safe by default. For situations in which
+    such protections are desirable, parallel safety can be enabled by setting
+    `parSafe = true` in any List constructor. A List object constructed from
+    another List object inherits the parallel safety mode of that List by
+    default.
+  */
+  record List {
+
+    /* The type of the elements contained in this List. */
+    type eltType;
+
+    /* If `true`, this List will perform parallel safe operations. */
+    param parSafe;
+
+    pragma "no doc"
+    var _size = 0;
+
+    //
+    // Little piece of wizardry I learned from @ronawho in #13104.
+    //
+    pragma "no doc"
+    var _lock$ = if parSafe then new LockWrapper() else none;
+
+    //
+    // Functionally an array of arrays.
+    //
+    pragma "no doc"
+    var _blocks: unmanaged ListBlock(unmanaged ListBlock(eltType)) = nil;
```

Comment:
I think it'd be less confusing if you just used _ddata here. It's OK for the normal allocator to run on the pointers to the blocks, since it'll just set them to `nil`. I also think it's probably more natural to manage the block indices internally as zero based. (In fact I'd view it that way also for the elements - you can subtract one once and for all in the list.this).

Reply:
In any case the actual problem with using `ListBlock` here is that `ListBlock` is a class. That results in an extra, unnecessary, indirection when doing an element access. I think you should just make `ListBlock` a record.

Reply:
Can you add a `typedef` for `_ddata`, or better yet standardize the usage of raw `_ddata`? A `_ddata` is just a C pointer with locality information attached, right? I know this is maybe for another issue, but if it was made more public, it'd be cool if it was called 'ptr` (to contrast with `c_ptr` but also demonstrate its similarity).

Reply:
If you want to weigh in on the names of `_ddata` we have an issue in which we are discussing that - #12490.

Reply:
I was thinking about the unnecessary indirection this morning, glad to know I thought correctly.